
<!DOCTYPE html
  PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head>
      <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
   <!--
This HTML was auto-generated from MATLAB code.
To make changes, update the MATLAB code and republish this document.
      --><title>ECES T580 Lab 5</title><meta name="generator" content="MATLAB 9.2"><link rel="schema.DC" href="http://purl.org/dc/elements/1.1/"><meta name="DC.date" content="2019-02-20"><meta name="DC.source" content="Lab5.m"><style type="text/css">
html,body,div,span,applet,object,iframe,h1,h2,h3,h4,h5,h6,p,blockquote,pre,a,abbr,acronym,address,big,cite,code,del,dfn,em,font,img,ins,kbd,q,s,samp,small,strike,strong,sub,sup,tt,var,b,u,i,center,dl,dt,dd,ol,ul,li,fieldset,form,label,legend,table,caption,tbody,tfoot,thead,tr,th,td{margin:0;padding:0;border:0;outline:0;font-size:100%;vertical-align:baseline;background:transparent}body{line-height:1}ol,ul{list-style:none}blockquote,q{quotes:none}blockquote:before,blockquote:after,q:before,q:after{content:'';content:none}:focus{outine:0}ins{text-decoration:none}del{text-decoration:line-through}table{border-collapse:collapse;border-spacing:0}

html { min-height:100%; margin-bottom:1px; }
html body { height:100%; margin:0px; font-family:Arial, Helvetica, sans-serif; font-size:10px; color:#000; line-height:140%; background:#fff none; overflow-y:scroll; }
html body td { vertical-align:top; text-align:left; }

h1 { padding:0px; margin:0px 0px 25px; font-family:Arial, Helvetica, sans-serif; font-size:1.5em; color:#d55000; line-height:100%; font-weight:normal; }
h2 { padding:0px; margin:0px 0px 8px; font-family:Arial, Helvetica, sans-serif; font-size:1.2em; color:#000; font-weight:bold; line-height:140%; border-bottom:1px solid #d6d4d4; display:block; }
h3 { padding:0px; margin:0px 0px 5px; font-family:Arial, Helvetica, sans-serif; font-size:1.1em; color:#000; font-weight:bold; line-height:140%; }

a { color:#005fce; text-decoration:none; }
a:hover { color:#005fce; text-decoration:underline; }
a:visited { color:#004aa0; text-decoration:none; }

p { padding:0px; margin:0px 0px 20px; }
img { padding:0px; margin:0px 0px 20px; border:none; }
p img, pre img, tt img, li img, h1 img, h2 img { margin-bottom:0px; } 

ul { padding:0px; margin:0px 0px 20px 23px; list-style:square; }
ul li { padding:0px; margin:0px 0px 7px 0px; }
ul li ul { padding:5px 0px 0px; margin:0px 0px 7px 23px; }
ul li ol li { list-style:decimal; }
ol { padding:0px; margin:0px 0px 20px 0px; list-style:decimal; }
ol li { padding:0px; margin:0px 0px 7px 23px; list-style-type:decimal; }
ol li ol { padding:5px 0px 0px; margin:0px 0px 7px 0px; }
ol li ol li { list-style-type:lower-alpha; }
ol li ul { padding-top:7px; }
ol li ul li { list-style:square; }

.content { font-size:1.2em; line-height:140%; padding: 20px; }

pre, code { font-size:12px; }
tt { font-size: 1.2em; }
pre { margin:0px 0px 20px; }
pre.codeinput { padding:10px; border:1px solid #d3d3d3; background:#f7f7f7; }
pre.codeoutput { padding:10px 11px; margin:0px 0px 20px; color:#4c4c4c; }
pre.error { color:red; }

@media print { pre.codeinput, pre.codeoutput { word-wrap:break-word; width:100%; } }

span.keyword { color:#0000FF }
span.comment { color:#228B22 }
span.string { color:#A020F0 }
span.untermstring { color:#B20000 }
span.syscmd { color:#B28C00 }

.footer { width:auto; padding:10px 0px; margin:25px 0px 0px; border-top:1px dotted #878787; font-size:0.8em; line-height:140%; font-style:italic; color:#878787; text-align:left; float:none; }
.footer p { margin:0px; }
.footer a { color:#878787; }
.footer a:hover { color:#878787; text-decoration:underline; }
.footer a:visited { color:#878787; }

table th { padding:7px 5px; text-align:left; vertical-align:middle; border: 1px solid #d6d4d4; font-weight:bold; }
table td { padding:7px 5px; text-align:left; vertical-align:top; border:1px solid #d6d4d4; }





  </style></head><body><div class="content"><h1>ECES T580 Lab 5</h1><!--introduction--><p>Bhautik (Brian) Amin</p><!--/introduction--><h2>Contents</h2><div><ul><li><a href="#1">Lab 5.1.1</a></li><li><a href="#7">Lab 5.2.1</a></li></ul></div><h2 id="1">Lab 5.1.1</h2><p><b><i>Background</i></b></p><p>SARS, a disease/illness that first appeared in China. It is a virus, and underwent mutations over a long period of time. The goal is to figure out what is the virus is related too</p><p>The pipeline consists of MAFFT, for quick alignment of multiple sequences with little computational burden. Some error but it works for our test case MAFFT will first align series of sequences, then we will pass the data onto readseq that will convert to a different data format (for running RAxML, "maximum likelihood"). From there we will take the data into iTOL where we will be able to visulize a tree</p><p>Here is the output from iTOL:</p><pre class="codeinput">figure(1)
imshow(<span class="string">'part1_tree.jpg'</span>)
</pre><img vspace="5" hspace="5" src="Lab5_01.png" alt=""> <p>Upon analysis we can see that Palm Civet is most likely the closest relative to Human SARS</p><h2 id="7">Lab 5.2.1</h2><p>Run SARS Data Part 2 through the same pipeline and show results</p><pre class="codeinput">figure(2)
imshow(<span class="string">'part2_tree.jpg'</span>)
</pre><img vspace="5" hspace="5" src="Lab5_02.png" alt=""> <p>Upon analysis we can see that the epidemic most likely originated from Zhongshan Guangdong in 12/26/02</p><p>Import the data sequence using fastaread</p><pre class="codeinput">[headers, sequences] = fastaread(<span class="string">'SARS_Data_Part_2.fasta'</span>);
</pre><p>Calculate pairwise distance between sequences</p><pre class="codeinput">D = seqpdist(sequences, <span class="string">'Alphabet'</span>, <span class="string">'NT'</span>,<span class="string">'indels'</span>,<span class="string">'pair'</span>);
</pre><p>Construct a phylogenetic tree from pairwise distances</p><pre class="codeinput">phyloTree = seqlinkage(D,<span class="string">'average'</span>,headers)
view(phyloTree)
</pre><pre class="codeoutput">    Phylogenetic tree object with 13 leaves (12 branches)
</pre><img vspace="5" hspace="5" src="Lab5_03.png" alt=""> <p>Upon analysis of this tree, we can see that the Zhongshan Guangdong 12/26/02 epidemic strain branched along with Guangzhou Guangdong 12/16/02 epidemic strain, and that they shared some common ancestor.</p><p>The difference between the RAxML model and the UPGMA model, in the UPGMA method, it makes the assumption that rate of mutations is constant over time,for all lineages in a tree. This means that all leaves of the tree have same distance from the root. Whereas, the RAxML shows sequences changing and mutating and varying length of times. Whereas the maximum likelihood model is seen to have root and branch distances varying, not utilizing this basic assumption of the molecular clock hypothesis</p><p class="footer"><br><a href="http://www.mathworks.com/products/matlab/">Published with MATLAB&reg; R2017a</a><br></p></div><!--
##### SOURCE BEGIN #####
%% ECES T580 Lab 5
% Bhautik (Brian) Amin



%% Lab 5.1.1

%% 
% *_Background_*
%%
% SARS, a disease/illness that first appeared in China. It is a virus, and
% underwent mutations over a long period of time. The goal is to figure out
% what is the virus is related too
%%
% The pipeline consists of MAFFT, for quick alignment of multiple sequences
% with little computational burden. Some error but it works for our test
% case
% MAFFT will first align series of sequences, then we will pass the data
% onto readseq that will convert to a different data format (for running RAxML,
% "maximum likelihood"). From there we will take the data into iTOL where
% we will be able to visulize a tree


%%
% Here is the output from iTOL:

figure(1)
imshow('part1_tree.jpg')

%%
% Upon analysis we can see that Palm Civet is most likely the closest
% relative to Human SARS


%% Lab 5.2.1

%%
% Run SARS Data Part 2 through the same pipeline and show results
figure(2)
imshow('part2_tree.jpg')

%%
% Upon analysis we can see that the epidemic most likely originated from
% Zhongshan Guangdong in 12/26/02


%%
% Import the data sequence using fastaread
[headers, sequences] = fastaread('SARS_Data_Part_2.fasta');

%%
% Calculate pairwise distance between sequences


D = seqpdist(sequences, 'Alphabet', 'NT','indels','pair');
%%
% Construct a phylogenetic tree from pairwise distances
phyloTree = seqlinkage(D,'average',headers)
view(phyloTree)

%%
% Upon analysis of this tree, we can see that the Zhongshan Guangdong
% 12/26/02 epidemic strain branched along with Guangzhou Guangdong 12/16/02
% epidemic strain, and that they shared some common ancestor. 

%%
% The difference between the RAxML model and the UPGMA model, in the UPGMA
% method, it makes the assumption that rate of mutations is constant over time,for
% all lineages in a tree. This means that all leaves of the tree have same 
% distance from the root. Whereas, the RAxML shows sequences changing and
% mutating and varying length of times. Whereas the maximum likelihood
% model is seen to have root and branch distances varying, not utilizing
% this basic assumption of the molecular clock hypothesis

##### SOURCE END #####
--></body></html>